### PR TITLE
Add a sample for testing databricks notebook with imported custom python libraries

### LINKED ISSUE #327: sample notebooks with python module, nutter/pytest UT and CI pipeline.

### PR DESCRIPTION
**Pull request by swetha-sundar**
# Type of PR

- Documentation changes
- Code changes
- Test changes
- CI-CD changes

## Purpose
- Add a sample for testing databricks notebook with imported custom python libraries
- Includes unit tests and integration tests
- Includes a devops workflow with both CI and CD pipelines based on Azure DevOps

## TODO/Pending
- Nutter based integration test notebook needs to be updated.

## Does this introduce a breaking change? If yes, details on what can break
No

## Author pre-publish checklist

- [x] Added test to prove my fix is effective or new feature works
- [x] No PII in logs
- [x] Made corresponding changes to the documentation

## Validation steps
- Refer to CI and CD Runs from this sample
- CI: https://dev.azure.com/OneCSEWeek/DatabricksOps/_build/results?buildId=491&view=results
- CD: https://dev.azure.com/OneCSEWeek/DatabricksOps/_build/results?buildId=493&view=results

## Issues Closed
- Closes #327

### CODE REVIEW

File: single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/ci-pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          pip install -r requirements.txt
+        displayName: 'Install libraries and dependencies'
+
+      - script: |
+          echo "y
+          $(databricksDomain)
+          $(databricksPat)
+          $(databricksClusterId)
+          $(databricksWorkspaceOrgId)
+          15001" | databricks-connect configure
```
Comment (namitms) [inline single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/ci-pipeline.yml:57]: What does 15001 represent?

File: single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+parameters:
+- name: environmentName
+  type: string
+- name: databricksDomain
+  type: string
+- name: databricksToken
+  type: string
+- name: databricksNotebookPath
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml:8]: Perhaps it makes more sense to call this `databricksLibsPath` ?

File: single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            addToPath: true
+            architecture: 'x64'
+          displayName: 'Use Python Version: $(pythonVersion)'
+
+        - script: |
+            python -m pip install --upgrade pip
+            pip install databricks-cli
+          displayName: 'Setup Agent'
+        - script: |
+            echo "Uploading libs at ${LIB_PATH} to workspace (${DATABRICKS_LIB_PATH})..."
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml:37]: You print out `"Uploading libs at ${LIB_PATH} to workspace (${DATABRICKS_LIB_PATH})..."`
But you actually upload to `dbfs:/FileStore/jars` :)

File: single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/ci-pipeline.yml (hunk 1 above)
Comment (swetha-sundar) [reply single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/ci-pipeline.yml:46]: It is the port number in which the databricks connection is configured.
Will convert that into a variable.

Comment (devlace) [reply single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/ci-pipeline.yml:46]: Its the port where databricks-connect connects to in the cluster. Maps to `spark.databricks.service.port`.

Comment (devlace) [reply single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/ci-pipeline.yml:46]: But yeah, perhaps add a comment saying its the Port ?

File: single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+pip
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/requirements.txt:1]: Why is this file here?

File: single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/README.md [hunk 5]
```diff
@@ -1 +1,49 @@
...
+- [Github account](https://github.com/) for source control
+- [Azure Databricks workspace](https://azure.microsoft.com/en-au/services/databricks/) with an [existing cluster](https://docs.microsoft.com/en-us/azure/databricks/clusters/create) running. Note: Refer to [Sample 1](../../sample1_basic_azure_databricks_environment/README.md) and [Sample 3](../../sample3_cluster_provisioning_and_data_access/README.md) to setup these resource.
+
+### Setup and deployment
+
+1. Clone (or import) this repository.
+1. In Azure DevOps, [define a Variable Group](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=yaml) called `mdwdo-dbx-npl-test` with the following variables:
+    1. **databricksDomain** - The databricks host in this format: `https://adb-123456789xxxx.x.azuredatabricks.net`
+    1. **databricksToken** - Databricks PAT token.
+    1. **databricksNotebookPath** - Where in the databricks workspace the notebook will be uploaded as part of the Release pipeline. ei. `/Shared/notebooks`
```
Comment (devlace) [inline single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/README.md:22]: Do we want to add `databricksLibsPath`? Or just hardcode for now to dbfs:/FileStore/jars?

**Review by devlace — CHANGES_REQUESTED** [review]
I tried out the pipeline and just have some minor comments :) 

- The linting is also failing. I recommend installing the markdownlint extension.
- We will also probably want to remove the .DS_Store files.

File: single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/requirements.txt (hunk 4 above)
Comment (swetha-sundar) [reply single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/requirements.txt:1]: @gary918 Are you using this file anywhere?  If you do, suggest combining with the requirements.txt in the parent folder "notebook-python-lib"

Comment (gary918) [reply single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/devops/requirements.txt:1]: @swetha-sundar No. I'm not. I'm Ok for combining them.

File: single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml (hunk 2 above)
Comment (gary918) [reply single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml]: @devlace Yes! It's been changed to databricksLibsPath.

File: single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml (hunk 3 above)
Comment (gary918) [reply single_tech_samples/databricks/sample4_ci_cd/devops/templates/jobs/deploy-libs-job.yml]: @devlace Changed it to use ${DATABRICKS_LIB_PATH}. Thanks :).

File: single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/README.md (hunk 5 above)
Comment (gary918) [reply single_tech_samples/databricks/sample4_ci_cd/notebook-python-lib/README.md:25]: @devlace Thanks. I've changed it to use databricksLibPath instead.

**Review by devlace — APPROVED** [review]
LGTM 👍

**Comment by gary918** [comment]
@swetha-sundar If databricksWorkspaceOrgId is still needed, we should also mention it in README.md?
